### PR TITLE
Add python files for agml predictor data preparation

### PR DESCRIPTION
**Pull request by umdsgy**
I have added several python files related to agml predictor data preparation, please review it.

### CODE REVIEW

**Comment by krsnapaudel** [comment]
@umdsgy: Please fix errors in script. See test failure details. Thanks.

**Comment by krsnapaudel** [comment]
@umdsgy : note that the python code produces no data for India. This needs investigation. R code works fine.

File: data_preparation/array.py [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+def arr_stats(arr, weights=None,
+              output=('min', 'max', 'sum', 'mean', 'count', 'mode')):
+    """
+    Extracts statistics from input array (arr).
+    It uses weights array as weights if provided.
+    List of statistics to extract is defined in the output parameter as a list or a comma separated string.
+    :param arr: Input array
+    :param weights: Array providing weights for each pixel value. Used to calculate stats.
+    :param output: List of values to extract, can be a list or comma separated string.
+        Possible values are:
```
Comment (pvojnov) [inline data_preparation/array.py:17]: Missing description for output option "mode"

File: data_preparation/array.py [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+        else:
+            ind = np.isnan(_arr)
+            out_vals['mean'] = np.ma.mean(_arr[~ind])
+
+    if 'std' in _output:
+        arr_size = np.size(arr_compressed)
+        if weights is not None:
+            # combine mask from the arr and create a compressed arr
+            weights_compressed = np.ma.array(weights, mask=_arr.mask).compressed()
+            ind = np.isnan(arr_compressed) | np.isnan(weights_compressed) | (_arr <= -9999)
```
Comment (pvojnov) [inline data_preparation/array.py:54]: These arrays might not be of the same shape, you are mixing compressed and uncompressed arrays

File: data_preparation/raster.py [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+                *args, **kwargs):
+    """
+    Reads the data fom the raster file and returns it as a numpy masked array.
+    The geometry is used to define the mask.
+    It returns only a subset of the data.
+        It reads only the raster window containing the geometry. This optimizes the performance by reducing disk reads.
+    The "use_pixels" parameter can be used to define how to handle the border pixels.
+        Some require shapely to be available on the system
+    :param ds: Raster file path or an instance of a rasterio.DatasetReader (already opened raster file).
+    :param mask: geojson geometry used to mask the array, mask is a read mask (only data covered by the geometry is used)
```
Comment (pvojnov) [inline data_preparation/raster.py:20]: The description should be updated to match the code change from the original. "mask" changed form geometry ->  iterable over geometries

**Review by pvojnov — COMMENTED** [review]
I see you didn't use our code for multiprocessing. But instead you are multiprocessing over multiple files.
Depending on hardware, number of spatial units and raster indicators this approach might run significantly slower since it reads multiple files at once and the disk IO might be a bottleneck.
If multiprocessing over geometries on each file you read one file at a time and extract multiple geometries at once, mostly CPU intense. That is why we used this approach in our code.

**Comment by krsnapaudel** [comment]
@pvojnov : Please check authorship as well. How you would like to be credited.

**Comment by umdsgy** [comment]
Please see the other PR

**Pull request closed without merging**